Add dependencies IP check

olexii4 opened the pull request:
Signed-off-by: Oleksii Orel <oorel@redhat.com>

<!-- Please review the following before submitting a PR:
Che's Contributing Guide: https://github.com/eclipse/che/blob/master/CONTRIBUTING.md
Pull Request Policy: https://github.com/eclipse/che/wiki/Development-Workflow#pull-requests

COMMITTERS: please include labels on each PR. Labels are listed here: https://github.com/eclipse/che/wiki/Labels but at a minimum you should include `kind/..` and `Dev Open Pull Request Status` labels.
-->

### What does this PR do?
Add dependencies IP check.

### What issues does this PR fix or reference?
fixes https://github.com/eclipse/che/issues/20058
<!-- #### Changelog -->
<!-- The changelog will be pulled from the PR's title. 
     Please provide a clear and meaningful title to the PR and don't include issue number -->


#### Release Notes
<!-- markdown to be included in marketing announcement - N/A for bugs -->


#### Docs PR
<!-- Please add a matching PR to [the docs repo](https://github.com/eclipse/che-docs) and link that PR to this issue.
Both will be merged at the same time. -->

sleshchenko commented on .deps/EXCLUDED/dev.md:
```diff
@@ -0,0 +1,33 @@
+This file contains a manual contribution to .deps/dev.md and it's needed because eclipse/dash-licenses does not deal with work-with CQs (more see https://github.com/eclipse/dash-licenses/issues/13)
```
Welcome back, automated dependencies analizying!

It would be great if we expose it from README.md, like:

### Dependencies IP

To make sure all the dependencies satisfy Eclipse Intellectual Property, this repo uses https://github.com/che-incubator/dash-licenses which is a wrapper for https://github.com/eclipse/dash-licenses.

So, check .deps/EXCLUDED/prod.md for dependencies use package and .deps/dev.md for ones we use at build time.

olexii4 replied:
I have fixed it.

sleshchenko commented on README.md:
```diff
@@ -70,20 +70,17 @@ To provide a custom remote server:
+this repo uses https://github.com/che-incubator/dash-licenses which is a wrapper for https://github.com/eclipse/dash-licenses.
 
-Firstly, build the license-tool dockerfile:
+So, check .deps/EXCLUDED/prod.md for dependencies use package and .deps/dev.md for ones we use at build time.
```
I know I provided it but just as an example ) It needs the following fixes:
```suggestion
So, check [.deps/prod.md](https://github.com/eclipse-che/che-dashboard/blob/main/.deps/prod.md) for dependencies we package and [.deps/dev.md](https://github.com/eclipse-che/che-dashboard/blob/main/.deps/dev.md) for ones we use at build time only.
```

olexii4 replied:
Ok. I have fixed it.

sleshchenko commented on README.md:
````diff
@@ -70,20 +70,17 @@ To provide a custom remote server:
-```
-
-and then run the license-tool:
+Run the license check:
````
It would be more clear if it says `To generate dependencies info`, since I don't expect from check run to generate anything.

olexii4 replied:
Ok. I have fixed it.

sleshchenko commented on .github/workflows/pr.yaml:
```diff
@@ -21,6 +21,16 @@ env:
+        uses: actions/checkout@v2
+      -
+        name: "Check dependencies usage restrictions"
+        run: docker run --rm -v ${PWD}:/workspace/project quay.io/che-incubator/dash-licenses:next --check
```
```suggestion
        run: yarn license:check
```

olexii4 replied:
Ok. I have fixed it.

sleshchenko reviewed (approved):
nothing critical but please consider fixing left comments

olexii4 commented on package.json:
```diff
@@ -28,7 +28,8 @@
     "test:coverage": "jest --coverage",
-    "test:watch": "jest --watch"
+    "test:watch": "jest --watch",
+    "licenseCheck": "docker run --rm -t -v ${PWD}/:/workspace/project quay.io/che-incubator/dash-licenses:next"
```
Ok. I have fixed it.

akurinnoy commented:
After checking out this branch and running `yarn run license:generate`

1. I got my `dev.md` and `prod.md` updated.

`prod.md`:
```diff
-| `@eclipse-che/api@7.32.1` | EPL-2.0 | clearlydefined |
+| `@eclipse-che/api@7.32.2` | EPL-2.0 |  |
```

`dev.md`:
```diff
-| [`electron-to-chromium@1.3.761`](https://github.com/kilian/electron-to-chromium/) | ISC | [CQ18717](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=18717) |
+| [`electron-to-chromium@1.3.762`](https://github.com/kilian/electron-to-chromium/) | ISC | [CQ18717](https://dev.eclipse.org/ipzilla/show_bug.cgi?id=18717) |
```

2. I also spotted that some of the excluded development dependencies are not handled correctly and treated as unresolved, e.g.:

https://github.com/eclipse-che/che-dashboard/blob/7ae46a95eed755a02271b104db334d269f786199/.deps/EXCLUDED/dev.md#L20

https://github.com/eclipse-che/che-dashboard/blob/7ae46a95eed755a02271b104db334d269f786199/.deps/dev.md#L689

The same is with `postcss-syntax@0.36.2` and `postcss-syntax@0.36.2`.

3. The list of exclusions of production dependencies should be revised and cleaned up.

olexii4 commented:
@akurinnoy 
1. These files 'dev.md' and 'prod.md' was generated before rebasing to the main branch. It brought some differences.
I have fixed it in [PR](https://github.com/che-incubator/dash-licenses/pull/4).
2.  Excluded development dependencies were not handled at all(a new registration which came with https://github.com/che-incubator/dash-licenses/commit/305e0d2ab25a89a546b6cb6e24112f898c08bea.
I have fixed it in [PR](https://github.com/che-incubator/dash-licenses/pull/4).
3.I have revised and cleaned up exclusions.

akurinnoy commented on README.md:
````diff
@@ -70,20 +70,17 @@ To provide a custom remote server:
 
 ```sh
-yarn licenseCheck:run
+yarn licenseCheck
````
```suggestion
yarn license:generate
```

olexii4 replied:
I have fixed it.

akurinnoy commented on .gitignore:
```diff
@@ -10,7 +10,7 @@ node_modules/
 yarn-error.log
 
-.deps/tmp/
+.deps/problems.md
```
I believe we should not ignore this file.

olexii4 replied:
I have fixed it.

akurinnoy commented on .yarnrc:
```diff
@@ -0,0 +1,5 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
I seem this file is a leftover you've forgotten to clean up, same with `.yarn/releases/yarn-1.22.5.cjs`.

olexii4 replied:
I have fixed it.
